### PR TITLE
Homework 2

### PR DESCRIPTION
**Pull request by asish918**
# Homework 2 Asish

### Topics Covered:
- Constructors
- Control Flow Statements
- Packages
- A few OOPs concepts

### Drift Video Explanation
[Drift](https://video.drift.com/v/ab0khEh4n89lb1RIJQgi5GxIiwtVM6zymYgzy0YvLk)

### Output Screenshots - 

Screenshot 2024-01-09 at 23 38 46
Screenshot 2024-01-09 at 23 38 20
Screenshot 2024-01-09 at 23 38 34

### CODE REVIEW

**Comment by asish918** [comment]
| Metric | Value |
| ------ | ----- |
| alert_status | <b>:white_check_mark: OK</b> |
| bugs | 0 |
| code_smells | 1 |
| reliability_rating | A (0 Bugs)<br/>Note: A being best, E being worst |
| security_rating | A (0 Vulnerabilities)<br/>Note: A being best, E being worst |
| vulnerabilities | 0 |

File: backend/homework/1/.gitignore [hunk 1]
```diff
@@ -0,0 +1,38 @@
+target/
```
Comment (rishav-kd) [inline backend/homework/1/.gitignore:1]: Dont commit homework 1 files

File: backend/homework/2/src/main/java/asish/kdu/logging/CustomLogger.java [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+import org.slf4j.LoggerFactory;
+
+public class CustomLogger {
+    private CustomLogger() {
+
+    }
+    private static final Logger logger
+            = LoggerFactory.getLogger(CustomLogger.class);
+    public static void printLogger(String message) {
+        logger.info(message);
```
Comment (rishav-kd) [inline backend/homework/2/src/main/java/asish/kdu/logging/CustomLogger.java:13]: use other logger levels for exceptions debugging, read a bit on this

File: backend/homework/2/src/main/java/asish/kdu/q1/Q1Utils.java [hunk 3]
```diff
@@ -0,0 +1,23 @@
+package asish.kdu.q1;
+
+import asish.kdu.logging.CustomLogger;
+
+public class Q1Utils {
+    private Q1Utils() {
+
```
Comment (rishav-kd) [inline backend/homework/2/src/main/java/asish/kdu/q1/Q1Utils.java:7]: Why are you using a private constructor here? Do you want to implement singleton class?

File: backend/homework/2/src/main/java/asish/kdu/q1/StudentUtil.java [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+public class StudentUtil {
+    public static double[] calculateGPA(int[] studentIdList, char[][]
+            studentsGrades) {
+
+        HashMap<Character, Integer> gradeMap = new HashMap<>();
+
+        gradeMap.put('A', 4);
+        gradeMap.put('B', 3);
+        gradeMap.put('C', 2);
+
```
Comment (rishav-kd) [inline backend/homework/2/src/main/java/asish/kdu/q1/StudentUtil.java:14]: Pls put this in a constructor, never do it outside a method.

File: backend/homework/2/src/main/java/asish/kdu/q1/StudentUtil.java [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+        return gpaResult;
+    }
+
+    public static int[] getStudentsByGPA(double lower, double higher, int[]
+            studentIdList, char[][] studentsGrades) {
+        if(lower < 0 || higher < 0 || lower > 4 || higher > 4 || higher < lower)
+            return new int[0];
+
+        double[] gpaResult = calculateGPA(studentIdList, studentsGrades);
```
Comment (rishav-kd) [inline backend/homework/2/src/main/java/asish/kdu/q1/StudentUtil.java:41]: Lets use dynamic arrays whenever possible.

File: backend/homework/2/src/main/java/asish/kdu/q2/Q2Utils.java [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+
+public class Q2Utils {
+    private Q2Utils() {
+
+    }
+
+    /**
+     * We define Opinion enum to re-use the same function for different scenarios
+     */
+    public enum Opinion {
```
Comment (rishav-kd) [inline backend/homework/2/src/main/java/asish/kdu/q2/Q2Utils.java:11]: Good usage!

File: backend/homework/2/src/main/java/asish/kdu/q3/APIResponseParser.java [hunk 7]
```diff
@@ -0,0 +1,62 @@
+package asish.kdu.q3;
+
+import asish.kdu.logging.CustomLogger;
+
+import java.util.Objects;
+
+public class APIResponseParser {
+    public static void main(String[] args) {
+        String xmlData =
+                "<work>\n<id type=\"integer\">2361393</id><books_count type=\"integer\">813</books_count>\n<ratings_count type=\"integer\">1,16,315</ratings_count>\n<text_reviews_count type=\"integer\">3439</text_reviews_count><original_publication_year type=\"integer\">1854</original_publication_year><original_publication_month type=\"integer\" nil=\"true\"/><original_publication_day type=\"integer\" nil=\"true\"/><average_rating>5.0</average_rating>\n<best_book type=\"Book\">\n<id type=\"integer\">16902</id> <title>Avengers</title> <author>\n<id type=\"integer\">10264</id>\n <name>Tony Stark</name> </author>\n <image_url>http://images.gr-assets.com/books/1465675526m/16902.jpg</image_url>\n <small_image_url>http://images.gr-assets.com/books/1465675526s/16902.jpg</small\n </small_image_url>\n </best_book>\n </work>";
```
Comment (rishav-kd) [inline backend/homework/2/src/main/java/asish/kdu/q3/APIResponseParser.java:10]: Lets keep this as a constant and save to a file `constants.java`

**Review by rishav-kd — COMMENTED** [review]
I liked the improvement from last PR, Only thing is use dynamic arrays going forward and  utilise different log levels for different logging situations

**Pull request closed without merging**